src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java

```diff
@@ -170,7 +172,15 @@ public void setApplicationContext(ApplicationContext applicationContext) throws
                 .read()
                 .flatMap(cosmosItemResponse -> Mono.justOrEmpty(toDomainObject(entityClass,
                     cosmosItemResponse.properties())))
-                .onErrorResume(this::databaseAccessExceptionHandler)
+                .onErrorResume(e -> {
+                    if (e instanceof CosmosClientException) {
+                        final CosmosClientException cosmosClientException = (CosmosClientException) e;
```

Comment:
Isn't it behavior regression?

Reply:
It may seem like, but it is actually not. This API broke the public contract of returning Optional empty result in case of document not found, so fixed it here. 
Regarding returning `CosmosClientException`, it makes more sense to return `CosmosClientException` instead of `DocumentDbAccessException` so that at least user can use the error message and status codes from the `CosmosClientException`. 

In future, I need to work on Exception handling in spring, and at that time, we can wrap `CosmosClientException` into some our own class of spring-data-cosmosdb exceptions, but for now, this is the least we can do.

Reply:
> It may seem like, but it is actually not. This API broke the public contract of returning Optional empty result in case of document not found, so fixed it here.

Is this spring requirement?

Reply:
Yes, it is. Spring APIs are meant be like that at least from a point of CrudRepository.java. 

They want find APIs to return empty or none if not found, and not error out. 

https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/CrudRepository.html#findById-ID-

So I followed the same contract, but the API didn't work as expected, so fixing it now :) 
```
/**
     * Retrieves an entity by its id.
     *
     * @param id must not be {@literal null}.
     * @param partitionKey partition key value of entity, must not be null.
     * @return the entity with the given id or {@literal Optional#empty()} if none found
     * @throws IllegalArgumentException if {@code id} is {@literal null}.
     */
    Optional<T> findById(ID id, PartitionKey partitionKey);
```

Reply:
> the entity with the given id or Optional#empty() if none found.

Its empty() optional or must?
If its optional then existing implementation also works. Changing such behaviors in a hotfix release is very big and risky. Can it be tracked for vNext?